---
title: "Escort"
author: "Xiaoru Dong"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Escort}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      warning = FALSE, 
                      error = FALSE, 
                      fig.align = "center", 
                      dpi = 300)
```


```{r, warning=F, message=F}
library(Escort)
set.seed(11111)
```

# Introduction

`Escort` is an R package designed to assist users in detecting trajectories, evaluating and ranking various analysis choices. It follows a three-step approach, guiding users through the analysis process by offering goodness-of-fit evaluations for embeddings that represent a range of analysis choices. This encompasses features like feature selection, dimension reduction, and trajectory inference methods, along with their respective hyperparameters.


# Run Escort
After successful installation, `Escort` and related packages need be loaded into the working space:

```{r, results='hide', message=FALSE, warning=FALSE}
library(Escort)
library(parallelDist) # for fast calculation of the distance matrix.
library(scuttle)
library(scran)
library(cluster)
library(mclust)
library(RMTstat)
```


# Data 
In this vignette, I'll be using a simulated scRNA-seq dataset featuring a linear trajectory structure.

```{r, warning=FALSE, error=FALSE, message=FALSE}
load("data/step0_clean_dyntoy_L3.RData")
```

Escort employs a three-step approach to guide users through the analysis by detecting the presence of trajectories and then offering goodness-of-fit evaluations for both embeddings and trajectories. 

## Step1: Detecting Trajectory Existence 

We identified two scenarios where trajectory fitting is not appropriate: when cells represent diverse cell types or when cells are homogeneous. The input in the step 1 is matrixes of raw and normalized single-cell RNA-seq data after quality control. 

### Identifing distinct cell types
```{r, warning=FALSE, error=FALSE, message=FALSE}
dist_mat <- parallelDist::parDist(t(norm_counts), method = "manhattan")
LvsC <- HD_DCClusterscheck(dist_mat=dist_mat, rawcounts=rawcounts)
LvsC$DCcheck
```

### Assessing the homogeneity 
```{r, warning=FALSE, error=FALSE, message=FALSE}
gene.var <- modelGeneVar(norm_counts)
HVGs <- getTopHVGs(gene.var, n=100)
cor_test <- testHomogeneous(HVGs=HVGs, norm_counts=norm_counts)
cor_test$decision
```

## Step2: Evaluating the Characteristics of Embeddings
The second step is designed to identify preferred embeddings for performing trajectory inference. 

We could select any number of highly variable genes for feature selection. Escort offers popular dimension reduction algorithms for generating embeddings, including UMAP, t-SNE, MDS, and PCA. In this case, we choose to apply the PCA algorithm, using the top 20% of highly variable genes for dimension reduction process.

```{r, warning=FALSE, error=FALSE, message=FALSE}
gene.var <- modelGeneVar(norm_counts)
genes.HVGs_1 <- getTopHVGs(gene.var, prop=0.2)
dimred <- getDR_2D(norm_counts, "PCA")
head(dimred)
```


### Examining cell connectivity on embeddings

We evaluate cell connectivity on the lower dimensional embedding. Since distinct clusters have already been eliminated in step 1, then a reliable embedding should not exhibit distinct clusters. 

```{r, warning=FALSE, error=FALSE, message=FALSE}
DRLvsC <- LD_DCClusterscheck(dist_mat=dist(dimred, method = "euclidean"), DRdims=dimred, connectedCells = 1)
DRLvsC$DCcheck
```



